### DISCUSSION TITLE
Arduino compatability

### DISCUSSION THREAD

**Discussion by fred-dev** [ssilverman/QNEthernet#53] [General]
The arduino page here: https://www.arduino.cc/reference/en/libraries/qnethernet/ 
Suggests: "This library is compatible with all architectures so you should be able to use it on all the Arduino boards."

I guess this is not true, but in case is there any way to use this with an ethernet shield on an Arduino Due? (I am using it for its 12 bit analogue input).

Cheers

**Comment by ssilverman** [comment]
Thank you for the question.

Unfortunately, the property that specifies this in the _library.properties_ file, `architectures`, isn't able to express the actual board. Teensy uses an ARM processor, however, it declares itself to the Arduino ecosystem as an AVR. I think this was because its core takes efforts to appear as an AVR-based Arduino for code that, say, uses AVR registers and things. This means that I can't use "arm" because of how the Teensy software declares itself, and I can't really use "avr" because my library is ARM-specific (the NXP M7 one that's on the Teensy).

Also, there's no way currently in Arduino-land to specify an actual board. For example, The Teensy has an ARM plus specific features, such as an Ethernet PHY. This means that, even if I was able to specify "arm" as the processor, there's no way to specify "teensy 4.1" specifically. In other words, the library properties file/specification is broken.

Side note: This is one of the reasons I made the library able to compile for non-Teensy 4.1 boards, because it is, technically, specified as `architecture=*`.

Hope this answers your question.

**Reply by fred-dev** [reply]
Hi, thanks so much for the detailed response, it all makes perfect sense.

**Comment by ssilverman** [comment]
@fred-dev I’ve recently made the library easier to compile for non-Teensy platforms and easier to make drivers for. (Watch for an upcoming push, probably later today or tomorrow.)

**Comment by ssilverman** [comment]
I recently pushed some changes that make it easier to write drivers for the library and run it on other platforms. I also included a W5500 driver, so it should, in theory, work with an Arduino Due. I don’t, however, have the hardware to test it. If you have the hardware setup, do you feel like trying it out?

**Comment by ssilverman** [comment]
I just tried to build for a Due, but the API is a little different. It doesn't yet work at this time.

**Reply by fred-dev** [reply]
Ok thanks. I do have the hardware so let me know if you want me to test it.

**Reply by ssilverman** [reply]
It looks like the `Print` API is missing some functions. As well, the library currently assumes a minimum of C++14.
